common/X11R5/clients/xclock/xclock.man
.\"ident	"@(#)r5xclock:xclock.man	1.1"
.TH XCLOCK 1 "Release 5" "X Version 11"
.SH NAME
xclock - analog / digital clock for X
.SH SYNOPSIS
.ta 8n
\fBxclock\fP	[-\fItoolkitoption\fP ...] [-help] [-analog] [-digital]
[-chime] [-hd \fIcolor\fP] 
.br
	[-hl \fIcolor\fP] [-update \fIseconds\fP] [-padding \fInumber\fP]
.SH DESCRIPTION
The
.I xclock 
program displays the time in analog or digital form.  The time is continuously
updated at a frequency which may be specified by the user.  This program is
nothing more than a wrapper around the Athena Clock widget.
.SH OPTIONS
.I Xclock
accepts all of the standard X Toolkit command line options along with the 
additional options listed below:
.TP 8
.B \-help
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
.TP 8
.B \-analog 
This option indicates that a conventional 12 hour clock face with tick marks
and hands should be used.  This is the default.
.TP 8
.B \-digital
This option indicates that a 24 hour digital clock should be used.
.TP 8
.B \-chime
This option indicates that the clock should chime 
once on the half hour and twice on the hour.
.TP 8
.B \-hd \fIcolor\fP
This option specifies the color of the hands on an analog clock.  The default
is \fIblack\fP.
.TP 8
.B \-hl \fIcolor\fP
This option specifies the color of the edges of the hands on an analog clock,
and is only useful on color displays.  The default is \fIblack\fP.
.TP 8
.B \-update \fIseconds\fP
This option specifies the frequency in seconds at which \fIxclock\fP
should update its display.  If the clock is obscured and then exposed,
it will be updated immediately.  A value of less than 30 seconds will enable a
second hand on an analog clock.  The default is 60 seconds.
.TP 8
.B \-padding \fInumber\fP
This option specifies the width in pixels of the padding 
between the window border and clock text or picture.  The default is 10
on a digital clock and 8 on an analog clock.
.SH X DEFAULTS
This program uses the 
.I Athena Clock
widget.  It understands all of the core resource names and classes as well as:
.PP
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the clock.  The default for analog clocks is 164
pixels; the default for digital clocks is whatever is needed to hold the 
clock when displayed in the chosen font.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the clock.  The default for analog clocks is 164
pixels; the default for digital clocks is whatever is needed to hold the
clock when displayed in the chosen font.
.TP 8
.B update (\fPclass\fB Interval)
Specifies the frequency in seconds at which the time should be redisplayed.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the tic marks. The default is depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP. 

.TP 8
.B hands (\fPclass\fB Foreground)
Specifies the color of the insides of the clock's hands. The default is 
depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP.
.TP 8
.B highlight (\fPclass\fB Foreground)
Specifies the color used to highlight the clock's hands. The default is
 depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP.
.TP 8
.B analog (\fPclass\fB Boolean)
Specifies whether or not an analog clock should be used instead of a digital
one.  The default is True.
.TP 8
.B chime (\fPclass\fB Boolean)
Specifies whether or not a bell should be rung on the hour and half hour.
.TP 8
.B padding (\fPclass\fB Margin)
Specifies the amount of internal padding in pixels to be used.  The default is
8.
.TP 8
.B font (\fPclass\fB Font)
Specifies the font to be used for the digital clock.  Note that variable width
fonts currently will not always display correctly.
.SH WIDGETS
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxclock\fR.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp
.nf
.TA .5i 
.ta .5i 
XClock  xclock
	Clock  clock
.fi
.sp
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH FILES
/usr/lib/X11/app-defaults/XClock - specifies required resources
.SH "SEE ALSO"
X(1), xrdb(1), time(3C), Athena Clock widget
.SH BUGS
.I Xclock
believes the system clock.
.PP
When in digital mode, the string should be centered automatically.
.SH COPYRIGHT
Copyright 1988, Massachusetts Institute of Technology.
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHORS
Tony Della Fera (MIT-Athena, DEC)
.br
Dave Mankins (MIT-Athena, BBN)
.br
Ed Moy (UC Berkeley)



common/X11R5/clients/xclock/xclock.c
/*	Copyright (c) 1990, 1991, 1992 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)r5xclock:xclock.c	1.1"
/* $XConsortium: xclock.c,v 1.32 91/04/19 13:46:20 converse Exp $ */

/*
 * xclock --  Hacked from Tony Della Fera's much hacked clock program.
 */

/*
 * Copyright 1989 Massachusetts Institute of Technology
 *
 *
 */

#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include <X11/Xaw/Clock.h>
#include <X11/Xaw/Cardinals.h>
#include "clock.bit"
#include "clmask.bit"

extern void exit();
static void quit();

/* Command line options table.  Only resources are entered here...there is a
   pass over the remaining options after XtParseCommand is let loose. */

static XrmOptionDescRec options[] = {
{"-chime",	"*clock.chime",		XrmoptionNoArg,		"TRUE"},
{"-hd",		"*clock.hands",		XrmoptionSepArg,	NULL},
{"-hands",	"*clock.hands",		XrmoptionSepArg,	NULL},
{"-hl",		"*clock.highlight",	XrmoptionSepArg,	NULL},
{"-highlight",	"*clock.highlight",	XrmoptionSepArg,	NULL},
{"-update",	"*clock.update",	XrmoptionSepArg,	NULL},
{"-padding",	"*clock.padding",	XrmoptionSepArg,	NULL},
{"-d",		"*clock.analog",	XrmoptionNoArg,		"FALSE"},
{"-digital",	"*clock.analog",	XrmoptionNoArg,		"FALSE"},
{"-analog",	"*clock.analog",	XrmoptionNoArg,		"TRUE"},
};


static XtActionsRec xclock_actions[] = {
    { "quit",	quit },
};

static Atom wm_delete_window;

/*
 * Report the syntax for calling xclock.
 */
Syntax(call)
	char *call;
{
	(void) printf ("Usage: %s [-analog] [-bw <pixels>] [-digital]\n", call);
	(void) printf ("       [-fg <color>] [-bg <color>] [-hd <color>]\n");
	(void) printf ("       [-hl <color>] [-bd <color>]\n");
	(void) printf ("       [-fn <font_name>] [-help] [-padding <pixels>]\n");
	(void) printf ("       [-rv] [-update <seconds>] [-display displayname]\n");
	(void) printf ("       [-geometry geom]\n\n");
	exit(1);
}

void main(argc, argv)
    int argc;
    char **argv;
{
    Widget toplevel;
    Arg arg;
    Pixmap icon_pixmap = None;
    XtAppContext app_con;

    toplevel = XtAppInitialize (&app_con, "XClock", options, XtNumber(options),
				&argc, argv, NULL, NULL, ZERO);
    if (argc != 1) Syntax(argv[0]);

    XtAppAddActions (app_con, xclock_actions, XtNumber(xclock_actions));

    /*
     * This is a hack so that f.delete will do something useful in this
     * single-window application.
     */
    XtOverrideTranslations(toplevel, 
		    XtParseTranslationTable ("<Message>WM_PROTOCOLS: quit()"));

    XtSetArg(arg, XtNiconPixmap, &icon_pixmap);
    XtGetValues(toplevel, &arg, ONE);
    if (icon_pixmap == None) {
	arg.value = (XtArgVal)XCreateBitmapFromData(XtDisplay(toplevel),
				       XtScreen(toplevel)->root,
				       (char *)clock_bits, clock_width, clock_height);
	XtSetValues (toplevel, &arg, ONE);
    }
    XtSetArg(arg, XtNiconMask, &icon_pixmap);
    XtGetValues(toplevel, &arg, ONE);
    if (icon_pixmap == None) {
	arg.value = (XtArgVal)XCreateBitmapFromData(XtDisplay(toplevel),
				       XtScreen(toplevel)->root,
				       (char *)clock_mask_bits, clock_mask_width, 
				       clock_mask_height);
	XtSetValues (toplevel, &arg, ONE);
    }

    XtCreateManagedWidget ("clock", clockWidgetClass, toplevel, NULL, ZERO);
    XtRealizeWidget (toplevel);
    wm_delete_window = XInternAtom (XtDisplay(toplevel), "WM_DELETE_WINDOW",
				    False);
    (void) XSetWMProtocols (XtDisplay(toplevel), XtWindow(toplevel),
			    &wm_delete_window, 1);
    XtAppMainLoop (app_con);
}


static void quit (w, event, params, num_params)
    Widget w;
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    if (event->type == ClientMessage &&
	event->xclient.data.l[0] != wm_delete_window) {
	XBell (XtDisplay(w), 0);
	return;
    }
    XCloseDisplay (XtDisplay(w));
    exit (0);
}
